internal/builders/nodejs/build.go

```diff
@@ -15,95 +15,226 @@
-	// 1. Retrieve the workflow inputs from env variable SLSA_WORKFLOW_INPUTS.
-
-	// 2. Install dependencies via `npm ci <inputs.ci-arguments>`
-
-	// 3. Build via `npm run <inputs.run-scripts>`
-
-	// 4. Create the final package tarball.
-	/* TODO: only run if non-empty.
-	   Note: pack-destination only supported version 7.x above.
-	   https://docs.npmjs.com/cli/v7/commands/npm-pack.
-	   This outputs a .tgz. Before running this command, let's record the .tgz
-	   files and their hashes, so that we can identify the new file without the need to parse
-	   the manifest.json.
-	   echo "npm pack --pack-destination="./out"
-	   copy tarball to upper folder to make the tarball accessible to next step.
-	*/
-	// TODO: output the list of artifacts and their corresponding build steps.
-	// The tarball name into a step output: echo "filename=$TARBALL" >> "$GITHUB_OUTPUT"
-
-	if dryRun {
-		// 1. Retrieve artifacts and their hases from env variable SLSA_BUILD_ARTIFACTS.
-
-		// 2. Output the proveance metadata in a format:
-		/* METADATA={
-			"provenance1.intoto.jsonl":{
-				"artifact-1":{
-					"digest": {
-						"sha256": "abcdef"
-					},
-					"buildSteps":[]Steps{
-						"workingDir": string,
-						"env": map[string]string,
-						"command": []string
-					}
-				},
-				"artifact-2":{
-					"digest": {
-						"sha256": "abcdef"
-					},
-					"buildSteps":[]Steps{
-						"workingDir": string,
-						"env": map[string]string,
-						"command": []string
-					}
-				}
-			},
-			"provenance2.intoto.jsonl":{
-				"artifact-3":{
-					"digest": {
-						"sha256": "abcdef"
-					},
-					"buildSteps":[]Steps{
-						"workingDir": string,
-						"env": map[string]string,
-						"command": []string
-					}
-				},
-			}
-		}
-		*/
+func build() error {
+	integration, err := slsaIntegrationNew()
+	if err != nil {
+		return err
+	}
+
+	r := runner.CommandRunner{}
+
+	// Ci command.
+	if err := createCiCommands(&r, integration.Inputs); err != nil {
 		return nil
 	}
 
-	// 4. Output the list of artifacts in a format TBD (sha256sum?).
+	// Run command.
+	if err := createRunCommands(&r, integration.Inputs); err != nil {
+		return err
+	}
+
+	// Pack command.
+	if err := createPackCommands(&r, integration.Inputs); err != nil {
+		return err
+	}
+
+	if integration.Inputs.DryRun {
+		return runDry(integration, r)
+	}
+
+	return run(integration, r)
+}
+
+func runDry(integration *slsaIntegration,
+	r runner.CommandRunner,
+) error {
+	// This builder supports a single provenance file generation.
+	if err := validateArtifacts(integration.Inputs.Artifacts); err != nil {
+		return nil
+	}
+	// Run dry.
+	steps, err := r.Dry()
+	if err != nil {
+		return err
+	}
+
+	// Generate the provenance metadata.
+	artifact := integration.Inputs.Artifacts[0]
+	name := strings.TrimSuffix(filepath.Base(artifact.Path), ".tgz")
+	metadata := slsaDryRunOutput{
+		name + ".intoto.jsonl": []slsaDryMetadata{
+			{
+				Name:    name,
+				Digests: artifact.Digests,
+				Steps:   steps,
```

Comment:
This is intentionally one possible format for the buildConfig right? i agree this is fine to start with, but i'm curious if people want to customize their own, since they're BYO-ing anyway

Reply:
good point. I hope we can have a recommended one by default, which is consistent with the buildType.The. we can update the version and let builders change the buildConfig / buildType.

I could rename Steps to buildConfig maybe? This way we can easily update in the next version by keeping the same format but allowing custom format. Wdut?

Note that all this code may become moot if the v1 specs only want the top-level inputs to be defined. If the specs go in this direction, we won't even need the dry run anymore: our reusable workflow will just provide the inputs it received.

Reply:
so I will make the buildConfig be a format the dry-run the dry run decide.